File: Back end/src/main/java/com/example/flowmerceproject/ProductManagement/service/CategoryService.java
package com.example.flowmerceproject.ProductManagement.service;

import com.example.flowmerceproject.ProductManagement.dto.CategoryDTOs;
import com.example.flowmerceproject.ProductManagement.entity.Category;
import com.example.flowmerceproject.ProductManagement.repository.CategoryRepository;
import com.example.flowmerceproject.StoreMangement.entity.Store;
import com.example.flowmerceproject.StoreMangement.repository.StoreRepository;
import com.example.flowmerceproject.UserManagement.exception.ConflictException;
import com.example.flowmerceproject.UserManagement.exception.ResourceNotFoundException;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class CategoryService {

    private final CategoryRepository categoryRepository;
    private final StoreRepository storeRepository;

    @Transactional
    public CategoryDTOs.CategoryResponse createCategory(CategoryDTOs.CategoryRequest request) {
        boolean nameExists = request.getStoreId() != null
                ? categoryRepository.existsByNameAndStore_StoreId(request.getName(), request.getStoreId())
                : categoryRepository.existsByName(request.getName());
        if (nameExists) {
            throw new ConflictException("Category already exists: " + request.getName());
        }
        Category category = Category.builder()
                .name(request.getName())
                .description(request.getDescription())
                .build();
        categoryRepository.save(category);
        return toResponse(category);
    }

    public List<CategoryDTOs.CategoryResponse> getAllCategories() {
        return categoryRepository.findAll()
                .stream().map(this::toResponse).collect(Collectors.toList());
    }

    public CategoryDTOs.CategoryResponse getCategoryById(Integer id) {
        return toResponse(categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found: " + id)));
    }

    @Transactional
    public CategoryDTOs.CategoryResponse updateCategory(Integer id,
                                                        CategoryDTOs.CategoryRequest request) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found: " + id));
        category.setName(request.getName());
        if (request.getDescription() != null) category.setDescription(request.getDescription());
        categoryRepository.save(category);
        return toResponse(category);
    }

    @Transactional
    public String deleteCategory(Integer id) {
        Category category = categoryRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found: " + id));
        categoryRepository.delete(category);
        return "Category deleted successfully.";
    }

    // ── STORE-SPECIFIC CATEGORY MANAGEMENT ───────────────────────────────────

    /** Returns global categories + this store's own categories, store-specific first. */
    public List<CategoryDTOs.CategoryResponse> getStoreCombinedCategories(Integer storeId) {
        List<Category> storeOwned = categoryRepository.findByStore_StoreId(storeId);
        List<Category> globals    = categoryRepository.findByStoreIsNull();

        List<CategoryDTOs.CategoryResponse> result = new ArrayList<>();
        Set<String> addedNames = new HashSet<>();

        for (Category c : storeOwned) {
            result.add(toResponse(c));
            addedNames.add(c.getName().toLowerCase());
        }
        for (Category c : globals) {
            if (!addedNames.contains(c.getName().toLowerCase())) {
                result.add(toResponse(c));
            }
        }
        return result;
    }

    /** Creates a category owned by a specific store (merchant-callable). */
    @Transactional
    public CategoryDTOs.CategoryResponse createStoreCategory(Integer storeId,
                                                              CategoryDTOs.CategoryRequest request) {
        if (categoryRepository.existsByNameAndStore_StoreId(request.getName(), storeId)) {
            throw new ConflictException("Category '" + request.getName() + "' already exists in your store.");
        }
        Store store = storeRepository.findById(storeId)
                .orElseThrow(() -> new ResourceNotFoundException("Store not found: " + storeId));
        Category category = Category.builder()
                .name(request.getName())
                .description(request.getDescription())
                .store(store)
                .build();
        categoryRepository.save(category);
        return toResponse(category);
    }

    /** Deletes a store-owned category (merchant can only delete their own). */
    @Transactional
    public void deleteStoreCategory(Integer storeId, Integer categoryId) {
        Category category = categoryRepository.findById(categoryId)
                .orElseThrow(() -> new ResourceNotFoundException("Category not found: " + categoryId));
        if (category.getStore() == null || !category.getStore().getStoreId().equals(storeId)) {
            throw new ConflictException("Cannot delete a global category.");
        }
        categoryRepository.delete(category);
    }

    public CategoryDTOs.CategoryResponse toResponse(Category c) {
        return CategoryDTOs.CategoryResponse.builder()
                .categoryId(c.getCategoryId())
                .storeId(c.getStore() != null ? c.getStore().getStoreId() : null)
                .name(c.getName())
                .description(c.getDescription())
                .build();
    }
}

File: frontend/services/category.service.ts
/**
 * Category service — wraps /categories endpoints.
 *
 * Backend endpoints:
 *   GET    /categories          → getAll()   (public)
 *   GET    /categories/{id}     → getById()  (public)
 *   POST   /categories          → create()   (ADMIN)
 *   PUT    /categories/{id}     → update()   (ADMIN)
 *   DELETE /categories/{id}     → delete()   (ADMIN)
 *
 * All methods return ApiResult<T>, never throw.
 */

import { httpClient } from '@/lib/api/http-client'
import type { ApiResult } from '@/types/api.types'
import type { CategoryRequest, CategoryResponse } from '@/types/category.types'

export const categoryService = {
  async getAll(): Promise<ApiResult<CategoryResponse[]>> {
    return httpClient.get<CategoryResponse[]>('/categories')
  },

  /** Returns global categories + this store's own categories combined (merchant auth). */
  async getStoreCategories(storeId: number, authHeaders?: Record<string, string>): Promise<ApiResult<CategoryResponse[]>> {
    return httpClient.get<CategoryResponse[]>(`/stores/${storeId}/categories`, authHeaders)
  },

  /** Creates a store-owned category (merchant auth). */
  async createStoreCategory(storeId: number, request: CategoryRequest, authHeaders?: Record<string, string>): Promise<ApiResult<CategoryResponse>> {
    return httpClient.post<CategoryResponse>(`/stores/${storeId}/categories`, request, authHeaders)
  },

  /** Deletes a store-owned category (merchant auth). */
  async deleteStoreCategory(storeId: number, categoryId: number, authHeaders?: Record<string, string>): Promise<ApiResult<void>> {
    return httpClient.delete<void>(`/stores/${storeId}/categories/${categoryId}`, authHeaders)
  },

  async getById(id: number): Promise<ApiResult<CategoryResponse>> {
    return httpClient.get<CategoryResponse>(`/categories/${id}`)
  },

  async create(request: CategoryRequest, authHeaders?: Record<string, string>): Promise<ApiResult<CategoryResponse>> {
    return httpClient.post<CategoryResponse>('/categories', request, authHeaders)
  },

  async update(id: number, request: CategoryRequest, authHeaders?: Record<string, string>): Promise<ApiResult<CategoryResponse>> {
    return httpClient.put<CategoryResponse>(`/categories/${id}`, request, authHeaders)
  },

  async delete(id: number, authHeaders?: Record<string, string>): Promise<ApiResult<void>> {
    return httpClient.delete<void>(`/categories/${id}`, authHeaders)
  },
}
